File: backend/src/services/GitScanService.ts
import simpleGit, { SimpleGit } from 'simple-git';
import * as R from 'ramda';
import { logger } from '../config/logger.js';

/** 提交时间 date 使用 author date，用于展示与统计的真实提交时间 */
export interface ScannedCommit {
  hash: string;
  message: string;
  date: number;
  authorName: string;
  authorEmail: string;
  filesChanged: number;
  insertions: number;
  deletions: number;
  branch?: string; // 提交所在的分支（可选，但按分支扫描时会填充）
}

export class GitScanService {
  private git: SimpleGit;

  constructor(private repoPath: string) {
    this.git = simpleGit(repoPath);
  }

  /**
   * 全图单次扫描：`git log --all` 跨所有 ref，每个 commit hash 至多出现一次（含 merge commit，不设 --no-merges）。
   * 不按分支写 `branch` 字段，入库多为 `branch=null`，与时间窗内多仓库汇总口径一致。
   */
  async scanRepositoryFlat(fromDate?: Date, toDate?: Date, authorEmails?: string[]): Promise<ScannedCommit[]> {
    const isRepo = await this.git.checkIsRepo();
    if (!isRepo) {
      throw new Error(`${this.repoPath} is not a valid git repository`);
    }

    const logOptions: string[] = ['--all'];
    if (fromDate) {
      logOptions.push('--since', fromDate.toISOString());
    }
    if (toDate) {
      logOptions.push('--until', toDate.toISOString());
    }
    if (authorEmails && authorEmails.length > 0) {
      const authorPattern = authorEmails
        .map(email => email.replace(/[.*+?^${}()[\]\\]/g, '\\$&'))
        .join('|');
      logOptions.push('--perl-regexp', '--author', authorPattern);
    }

    try {
      const logs = await this.git.log(logOptions);
      const commits = await Promise.all(
        R.map((commit) => this.mapLogEntryToScanned(commit, undefined), logs.all)
      );
      logger.info(`Flat scan (--all): ${commits.length} unique commits in ${this.repoPath}`);
      return commits;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      const errorStack = error instanceof Error ? error.stack : undefined;
      logger.error({
        msg: `Error flat-scanning repository ${this.repoPath}`,
        error: errorMessage,
        stack: errorStack,
        repoPath: this.repoPath
      });
      throw error;
    }
  }

  /**
   * 扫描仓库提交记录（按分支分别扫描；先扫未上线分支再扫 release，按 commitHash 去重且优先保留未上线分支名）
   * 保留作兼容；日常入库请使用 {@link scanRepositoryFlat}（由 {@link incrementalScan} 调用）。
   * @param fromDate 起始日期，用于增量扫描
   * @param toDate 结束日期，可选
   * @param authorEmails 作者邮箱数组，用于过滤（支持多个作者）
   */
  async scanRepository(fromDate?: Date, toDate?: Date, authorEmails?: string[]): Promise<ScannedCommit[]> {
    try {
      const isRepo = await this.git.checkIsRepo();
      if (!isRepo) {
        throw new Error(`${this.repoPath} is not a valid git repository`);
      }

      const { branchesToScan } = await this.getBranchesToScan();
      logger.info(`Found ${branchesToScan.length} branches to scan in ${this.repoPath}`);

      const allCommits: ScannedCommit[] = [];

      for (const branchName of branchesToScan) {
        try {
          const branchCommits = await this.scanBranch(branchName, fromDate, toDate, authorEmails);
          allCommits.push(...branchCommits);
          logger.debug(`Scanned ${branchCommits.length} commits from branch ${branchName}`);
        } catch (error) {
          const errorMessage = error instanceof Error ? error.message : String(error);
          logger.warn(`Failed to scan branch ${branchName}: ${errorMessage}`);
        }
      }

      // 按 commitHash 去重，同一 hash 只保留一条；优先保留带「未上线分支」名的条（branch 有值且非 release）
      const uniqueCommitsMap = new Map<string, ScannedCommit>();
      for (const commit of allCommits) {
        const key = commit.hash;
        const existing = uniqueCommitsMap.get(key);
        const currentIsUnreleased = commit.branch && commit.branch !== 'release' && commit.branch !== 'master';
        if (!existing) {
          uniqueCommitsMap.set(key, commit);
        } else if (currentIsUnreleased && (!existing.branch || existing.branch === 'release' || existing.branch === 'master')) {
          uniqueCommitsMap.set(key, commit);
        }
      }

      const uniqueCommits = Array.from(uniqueCommitsMap.values());
      logger.info(`Found ${uniqueCommits.length} unique commits (across ${branchesToScan.length} branches)`);

      return uniqueCommits;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      const errorStack = error instanceof Error ? error.stack : undefined;
      logger.error({
        msg: `Error scanning repository ${this.repoPath}`,
        error: errorMessage,
        stack: errorStack,
        repoPath: this.repoPath
      });
      throw error;
    }
  }

  /**
   * 获取待扫描分支列表：release 与 master 同时存在时以 release 为主（不扫 master）；
   * 只有 master 没有 release 时以 master 为主。非主分支若 tip 已在主分支中则跳过。顺序：未上线分支先，主分支最后。
   */
  private async getBranchesToScan(): Promise<{ releaseRef: string | null; branchesToScan: string[] }> {
    const raw = await this.getAllBranches();
    const releaseRef = raw.find(b => this.normalizeBranchNameForCompare(b) === 'release') ?? null;
    const masterRef = raw.find(b => this.normalizeBranchNameForCompare(b) === 'master') ?? null;

    const hasRelease = !!releaseRef;
    const hasMaster = !!masterRef;
    let list = raw;
    if (hasRelease && hasMaster) {
      list = raw.filter(b => this.normalizeBranchNameForCompare(b) !== 'master');
    }
    const mainRef = releaseRef ?? masterRef;
    const others = list.filter(b => b !== mainRef);

    const unreleased: string[] = [];
    for (const branch of others) {
      if (!mainRef) {
        unreleased.push(branch);
        continue;
      }
      try {
        const tip = await this.git.revparse([branch]);
        if (!tip || !tip.trim()) continue;
        const isAncestor = await this.git.raw(['merge-base', '--is-ancestor', tip.trim(), mainRef]).then(
          () => true,
          () => false
        );
        if (!isAncestor) {
          unreleased.push(branch);
        }
      } catch {
        unreleased.push(branch);
      }
    }

    const branchesToScan = [...unreleased];
    if (mainRef) branchesToScan.push(mainRef);
    return { releaseRef, branchesToScan };
  }

  /** 用于比较的分支名：去掉 origin/ 前缀，便于判断 release/master */
  private normalizeBranchNameForCompare(branchName: string): string {
    return branchName.replace(/^origin\//, '').trim();
  }

  /**
   * 获取所有分支（本地和远程）
   * 返回格式：本地分支直接返回名称，远程分支返回 origin/branchName 格式
   */
  private async getAllBranches(): Promise<string[]> {
    try {
      const localBranches = await this.git.branchLocal();
      const remoteBranches = await this.git.branch(['-r']);
      const branches = new Set<string>();

      for (const branch of localBranches.all) {
        if (!branch.includes('HEAD') && !branch.includes('->')) {
          branches.add(branch);
        }
      }
      for (const branch of remoteBranches.all) {
        if (!branch.includes('HEAD') && !branch.includes('->')) {
          const normalized = branch.replace(/^remotes\//, '');
          branches.add(normalized);
        }
      }
      return Array.from(branches);
    } catch (error) {
      logger.warn(`Failed to get all branches: ${error}`);
      try {
        const currentBranch = await this.git.branch();
        return currentBranch.current ? [currentBranch.current] : [];
      } catch {
        return [];
      }
    }
  }

  /**
   * 扫描单个分支的提交记录
   */
  private async scanBranch(
    branchName: string,
    fromDate?: Date,
    toDate?: Date,
    authorEmails?: string[]
  ): Promise<ScannedCommit[]> {
    // 确定实际的分支名称和存储用的分支名称
    // branchName 可能是 "origin/feature/xxx" 或 "feature/xxx"
    let actualBranchName = branchName; // 用于 git.log() 的分支名称
    let storageBranchName = branchName; // 用于存储到数据库的分支名称
    
    // 如果分支名称以 origin/ 开头，说明是远程分支
    if (branchName.startsWith('origin/')) {
      actualBranchName = branchName; // 直接使用 origin/xxx 格式扫描
      storageBranchName = branchName.replace(/^origin\//, ''); // 存储时去掉 origin/ 前缀
    } else {
      // 本地分支，直接使用
      actualBranchName = branchName;
      storageBranchName = branchName;
    }
    
    try {

      // 构建 log 选项数组
      const logOptions: string[] = [actualBranchName]; // 使用实际的分支名称

      if (fromDate) {
        logOptions.push('--since', fromDate.toISOString());
      }

      if (toDate) {
        logOptions.push('--until', toDate.toISOString());
      }

      if (authorEmails && authorEmails.length > 0) {
        const authorPattern = authorEmails.map(email =>
          email.replace(/[.*+?^${}()[\]\\]/g, '\\$&')
        ).join('|');
        logOptions.push('--perl-regexp');
        logOptions.push('--author', authorPattern);
      }

      // 获取该分支的提交日志
      const logs = await this.git.log(logOptions);

      const commits = await Promise.all(
        R.map((commit) => this.mapLogEntryToScanned(commit, storageBranchName || undefined), logs.all)
      );

      return commits;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      // 如果扫描失败，记录警告（使用 warn 级别以便排查问题）
      // 注意：actualBranchName 可能未定义（如果错误发生在初始化之前）
      const actualName = typeof actualBranchName !== 'undefined' ? actualBranchName : branchName;
      logger.warn(`Failed to scan branch ${branchName} (actual: ${actualName}): ${errorMessage}`);
      return [];
    }
  }

  /**
   * 将 simple-git 单条 log 转为 ScannedCommit（含 diff 统计；根提交无父时 stats 为 0）
   */
  private async mapLogEntryToScanned(
    commit: { hash: string; message: string; date: string; author_name: string; author_email: string },
    branch: string | undefined
  ): Promise<ScannedCommit> {
    try {
      const diff = await this.git.diffSummary([`${commit.hash}^`, commit.hash]);
      return {
        hash: commit.hash,
        message: commit.message,
        date: new Date(commit.date).getTime(),
        authorName: commit.author_name,
        authorEmail: commit.author_email,
        filesChanged: diff.files.length,
        insertions: diff.insertions,
        deletions: diff.deletions,
        branch
      };
    } catch (error) {
      logger.debug(`Failed to get diff for commit ${commit.hash}: ${error}`);
      return {
        hash: commit.hash,
        message: commit.message,
        date: new Date(commit.date).getTime(),
        authorName: commit.author_name,
        authorEmail: commit.author_email,
        filesChanged: 0,
        insertions: 0,
        deletions: 0,
        branch
      };
    }
  }

  /**
   * 增量扫描：只扫描指定日期之后的提交（全图 `--all`，含 merge）
   * @param lastScanDate 起始日期
   * @param authorEmails 作者邮箱数组，可选（支持多个作者）
   * @param toDate 结束日期，可选
   */
  async incrementalScan(lastScanDate: Date, authorEmails?: string[], toDate?: Date): Promise<ScannedCommit[]> {
    logger.info(`Incremental flat scan from ${lastScanDate.toISOString()}${toDate ? ` to ${toDate.toISOString()}` : ''}`);
    return this.scanRepositoryFlat(lastScanDate, toDate, authorEmails);
  }
}

File: python-backend/app/services/git_scan_service.py
"""Git scan service – GitPython port of GitScanService.ts."""

from __future__ import annotations

import asyncio
import logging
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class ScannedCommit:
    hash: str
    message: str
    date: int  # milliseconds
    authorName: str
    authorEmail: str
    filesChanged: int = 0
    insertions: int = 0
    deletions: int = 0
    branch: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "hash": self.hash,
            "message": self.message,
            "date": self.date,
            "authorName": self.authorName,
            "authorEmail": self.authorEmail,
            "filesChanged": self.filesChanged,
            "insertions": self.insertions,
            "deletions": self.deletions,
        }
        if self.branch is not None:
            d["branch"] = self.branch
        return d


class GitScanService:
    """Scan a local git repository for commits using GitPython.

    All potentially blocking git operations are wrapped in
    ``asyncio.to_thread`` so they can be awaited from async code.
    """

    def __init__(self, repo_path: str) -> None:
        self.repo_path = repo_path

    # ------------------------------------------------------------------
    # internal: open repo (non-blocking wrapper)
    # ------------------------------------------------------------------

    def _open_repo(self) -> "Repo":  # noqa: F821
        from git import Repo

        return Repo(self.repo_path)

    # ------------------------------------------------------------------
    # scan_repository_flat  (--all, cross-ref, single pass)
    # ------------------------------------------------------------------

    async def scan_repository_flat(
        self,
        from_date: datetime | None = None,
        to_date: datetime | None = None,
        author_emails: list[str] | None = None,
    ) -> list[ScannedCommit]:
        return await asyncio.to_thread(
            self._scan_repository_flat_sync,
            from_date,
            to_date,
            author_emails,
        )

    def _scan_repository_flat_sync(
        self,
        from_date: datetime | None,
        to_date: datetime | None,
        author_emails: list[str] | None,
    ) -> list[ScannedCommit]:
        from git import Repo

        repo = Repo(self.repo_path)

        # build git-log arguments
        args = ["--all", "--format=%H|%s|%aI|%an|%ae"]
        if from_date is not None:
            args.extend(["--since", from_date.isoformat()])
        if to_date is not None:
            args.extend(["--until", to_date.isoformat()])
        if author_emails and len(author_emails) > 0:
            pattern = "|".join(
                re.escape(e) for e in author_emails
            )
            args.extend(["--perl-regexp", "--author", pattern])

        raw_output: str = repo.git.log(*args)
        if not raw_output.strip():
            return []

        lines = raw_output.strip().split("\n")
        commits: list[ScannedCommit] = []
        for line in lines:
            parts = line.split("|", 4)
            if len(parts) < 5:
                continue
            commit_hash, message, date_iso, author_name, author_email = parts
            try:
                dt = datetime.fromisoformat(date_iso)
                date_ms = int(dt.timestamp() * 1000)
            except (ValueError, TypeError):
                continue

            # get diff stats
            files_changed = 0
            insertions = 0
            deletions = 0
            try:
                stats = repo.commit(commit_hash).stats
                files_changed = stats.files_changed if hasattr(stats, "files_changed") else len(stats.files)
                insertions = stats.total.get("insertions", 0) if isinstance(stats.total, dict) else stats.insertions
                deletions = stats.total.get("deletions", 0) if isinstance(stats.total, dict) else stats.deletions
            except Exception:
                pass  # root commit or other issue – defaults to 0

            commits.append(
                ScannedCommit(
                    hash=commit_hash,
                    message=message,
                    date=date_ms,
                    authorName=author_name,
                    authorEmail=author_email,
                    filesChanged=files_changed,
                    insertions=insertions,
                    deletions=deletions,
                )
            )

        logger.info(
            "Flat scan (--all): %d unique commits in %s",
            len(commits),
            self.repo_path,
        )
        return commits

    # ------------------------------------------------------------------
    # scan_repository  (branch-aware, legacy compatibility)
    # ------------------------------------------------------------------

    async def scan_repository(
        self,
        from_date: datetime | None = None,
        to_date: datetime | None = None,
        author_emails: list[str] | None = None,
    ) -> list[ScannedCommit]:
        return await asyncio.to_thread(
            self._scan_repository_sync,
            from_date,
            to_date,
            author_emails,
        )

    def _scan_repository_sync(
        self,
        from_date: datetime | None,
        to_date: datetime | None,
        author_emails: list[str] | None,
    ) -> list[ScannedCommit]:
        from git import Repo

        repo = Repo(self.repo_path)
        branches_to_scan = self._get_branches_to_scan(repo)
        logger.info(
            "Found %d branches to scan in %s",
            len(branches_to_scan),
            self.repo_path,
        )

        all_commits: list[ScannedCommit] = []
        for branch_name in branches_to_scan:
            try:
                branch_commits = self._scan_branch(
                    repo, branch_name, from_date, to_date, author_emails
                )
                all_commits.extend(branch_commits)
            except Exception as exc:
                logger.warning("Failed to scan branch %s: %s", branch_name, exc)

        # dedup by hash, prefer unreleased branch names
        unique: dict[str, ScannedCommit] = {}
        for commit in all_commits:
            key = commit.hash
            existing = unique.get(key)
            current_is_unreleased = (
                commit.branch is not None
                and commit.branch not in ("release", "master")
            )
            if existing is None:
                unique[key] = commit
            elif current_is_unreleased and (
                existing.branch is None
                or existing.branch in ("release", "master")
            ):
                unique[key] = commit

        result = list(unique.values())
        logger.info(
            "Found %d unique commits (across %d branches)",
            len(result),
            len(branches_to_scan),
        )
        return result

    # ------------------------------------------------------------------
    # incremental_scan  (delegates to flat scan)
    # ------------------------------------------------------------------

    async def incremental_scan(
        self,
        last_scan_date: datetime,
        author_emails: list[str] | None = None,
        to_date: datetime | None = None,
    ) -> list[ScannedCommit]:
        logger.info(
            "Incremental flat scan from %s%s",
            last_scan_date.isoformat(),
            f" to {to_date.isoformat()}" if to_date else "",
        )
        return await self.scan_repository_flat(last_scan_date, to_date, author_emails)

    # ------------------------------------------------------------------
    # internal: branch helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _normalize_branch_for_compare(name: str) -> str:
        return re.sub(r"^origin/", "", name).strip()

    def _get_branches_to_scan(self, repo: "Repo") -> list[str]:  # noqa: F821
        raw = self._get_all_branches(repo)
        release_ref = next(
            (b for b in raw if self._normalize_branch_for_compare(b) == "release"),
            None,
        )
        master_ref = next(
            (b for b in raw if self._normalize_branch_for_compare(b) == "master"),
            None,
        )

        has_release = release_ref is not None
        has_master = master_ref is not None
        lst = list(raw)
        if has_release and has_master:
            lst = [b for b in lst if self._normalize_branch_for_compare(b) != "master"]

        main_ref = release_ref or master_ref
        others = [b for b in lst if b != main_ref]

        unreleased: list[str] = []
        for branch in others:
            if not main_ref:
                unreleased.append(branch)
                continue
            try:
                tip = repo.git.rev_parse(branch).strip()
                if not tip:
                    continue
                # check if tip is ancestor of main
                repo.git.merge_base("--is-ancestor", tip, main_ref)
                # if we get here, tip IS an ancestor -> skip
            except Exception:
                unreleased.append(branch)

        result = list(unreleased)
        if main_ref:
            result.append(main_ref)
        return result

    @staticmethod
    def _get_all_branches(repo: "Repo") -> list[str]:  # noqa: F821
        branches: set[str] = set()
        try:
            for ref in repo.branches:
                name = str(ref)
                if "HEAD" not in name and "->" not in name:
                    branches.add(name)
            for ref in repo.remotes.origin.refs:
                name = str(ref).replace("remotes/", "")
                if "HEAD" not in name and "->" not in name:
                    branches.add(name)
        except Exception:
            try:
                if repo.active_branch:
                    branches.add(str(repo.active_branch))
            except Exception:
                pass
        return list(branches)

    def _scan_branch(
        self,
        repo: "Repo",  # noqa: F821
        branch_name: str,
        from_date: datetime | None,
        to_date: datetime | None,
        author_emails: list[str] | None,
    ) -> list[ScannedCommit]:
        # determine actual vs storage branch name
        actual_branch = branch_name
        storage_branch = branch_name
        if branch_name.startswith("origin/"):
            storage_branch = branch_name[len("origin/"):]

        args = [actual_branch, "--format=%H|%s|%aI|%an|%ae"]
        if from_date is not None:
            args.extend(["--since", from_date.isoformat()])
        if to_date is not None:
            args.extend(["--until", to_date.isoformat()])
        if author_emails and len(author_emails) > 0:
            pattern = "|".join(re.escape(e) for e in author_emails)
            args.extend(["--perl-regexp", "--author", pattern])

        raw_output: str = repo.git.log(*args)
        if not raw_output.strip():
            return []

        commits: list[ScannedCommit] = []
        for line in raw_output.strip().split("\n"):
            parts = line.split("|", 4)
            if len(parts) < 5:
                continue
            commit_hash, message, date_iso, author_name, author_email = parts
            try:
                dt = datetime.fromisoformat(date_iso)
                date_ms = int(dt.timestamp() * 1000)
            except (ValueError, TypeError):
                continue

            files_changed = 0
            insertions = 0
            deletions = 0
            try:
                stats = repo.commit(commit_hash).stats
                files_changed = len(stats.files)
                total = stats.total
                insertions = total.get("insertions", 0) if isinstance(total, dict) else 0
                deletions = total.get("deletions", 0) if isinstance(total, dict) else 0
            except Exception:
                pass

            commits.append(
                ScannedCommit(
                    hash=commit_hash,
                    message=message,
                    date=date_ms,
                    authorName=author_name,
                    authorEmail=author_email,
                    filesChanged=files_changed,
                    insertions=insertions,
                    deletions=deletions,
                    branch=storage_branch or None,
                )
            )
        return commits
